lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/REEFEnvironment.java

```diff
@@ -0,0 +1,154 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.runtime.common;
+
+import org.apache.reef.runtime.common.launch.ProfilingStopHandler;
+import org.apache.reef.runtime.common.launch.REEFErrorHandler;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Injector;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+import org.apache.reef.tang.exceptions.InjectionException;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+import org.apache.reef.util.EnvironmentUtils;
+import org.apache.reef.util.REEFVersion;
+import org.apache.reef.wake.profiler.WakeProfiler;
+import org.apache.reef.wake.time.Clock;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The main entry point into any REEF process (Driver and Evaluator).
+ * It is mostly reading from the command line to instantiate
+ * the runtime clock and calling .run() on it.
+ */
+public final class REEFEnvironment implements Runnable, AutoCloseable {
+
+  /**
+   * Parameter to enable Wake network profiling. By default profiling is disabled.
+   */
+  @NamedParameter(doc = "If true, profiling will be enabled", short_name = "profiling", default_value = "false")
+  private static final class ProfilingEnabled implements Name<Boolean> { }
+
+  private static final Logger LOG = Logger.getLogger(REEFEnvironment.class.getName());
+
+  private static final Tang TANG = Tang.Factory.getTang();
+
+  /** Main event loop of current REEF component (Driver or Evaluator). */
+  private final Clock clock;
+
+  /** Error handler that processes all uncaught REEF exceptions. */
+  private final REEFErrorHandler errorHandler;
+
+  /**
+   * Create new REEF environment.
+   * @param config REEF component (Driver or Evaluator) configuration.
+   * It is usually read from config file by REEFLauncher.
+   * @throws InjectionException Thrown on configuration error.
+   */
+  @SuppressWarnings("checkstyle:illegalcatch")
+  public REEFEnvironment(final Configuration config) throws InjectionException {
+
+    if (LOG.isLoggable(Level.FINEST)) {
+      LOG.log(Level.FINEST, "Configuration:\n--\n{0}\n--",
+          new AvroConfigurationSerializer().toString(config, true));
```

Comment:
1. this is pre-existing code, moved here from `REEFLauncher`;
   2. there is a chicken and egg problem - at this point we don't have an injector yet, and creating one can (theoretically) raise an `IllegalStateException`. Current code tries to log the configuration before building an injector with it.
   3. Instead of injecting a config serializer, we can do config pretty printing more elegantly inside the injector, exposing `Injector.toString()` method to the outside.

Either way, I would probably do it in a separate PR